### ISSUE TITLE
Panic caused by sparkline issue

### ISSUE THREAD

**Issue by chneau** [cjbassi/gotop#12]
I got it 5 times today. It generally happen when I do some heavy cpu docker build.

After the panic, the terminal is unusable, I have to close it and restart it.

If you need more informations on the system I'll be happy to help.

![image](https://user-images.githubusercontent.com/5753629/38506056-b6f39570-3c10-11e8-9ca3-291dc8cbd1d6.png)

**Comment by cjbassi** [comment]
Yah could you tell me what your OS, terminal, and CPU arch are? Also I added some logging so if you want you could `go get github.com/cjbassi/gotop, run gotop, and then let me know what happens or what the error messages are when they come up again. Thanks!

**Comment by barn3y** [comment]
I'm seeing this to, it's challenging to get a proper stacktrace, since the terminal is non-responsive after a crash, I can't even access the scrollback buffer.

I'm on Ubuntu 17.04, using tilix 1.5.6. Arch is amd64.

Here's a screenshot of the terminal with the end of the stacktrace, if it helps.

![screenshot from 2018-04-10 17-03-08](https://user-images.githubusercontent.com/7979/38541001-18c57efa-3ce1-11e8-9ab2-a2897bea113c.png)

**Comment by chneau** [comment]
Here is some details:
```
16:50:57-c@c:~/save$uname -a
Linux c 4.13.0-37-generic #42~16.04.1-Ubuntu SMP Wed Mar 7 16:03:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

```
It's an Ubuntu 16.04 using the default terminal which I is gnome terminal. 
Here is a new screenshot with the updated code:

![image](https://user-images.githubusercontent.com/5753629/38541249-6aff822e-3c96-11e8-954a-26b1f7dd839c.png)
Notice this:
```
index: -3
len(SPARKS): 8 
```
I did a ```go get -u github.com/cjbassi/gotop``` and ``` go install github.com/cjbassi/gotop ``` then used the gotop command on my home (not sure if it helps but too much information is always good).

And to trigger this error, my best method is to start a heavy Docker build (basically a npm install in a Docker, didn't try without docker tho).

As @barn3y said, I can't copy past the error, so here is a screenshot.

Here is another:

![image](https://user-images.githubusercontent.com/5753629/38541518-3d433dca-3c97-11e8-8fd6-09d6f2c9858f.png)

Basically, if you have a nodejs project and run npm install, it crashes. 

Here is my package.json.
```
{
  "name": "test",
  "private": true,
  "scripts": {
    "start": "meteor run"
  },
  "dependencies": {
    "@babel/runtime": "^7.0.0-beta.44",
    "babel-runtime": "^6.26.0",
    "dygraphs": "^2.1.0",
    "dygraphs-synchronizer": "^1.0.9",
    "leaflet": "^1.3.1",
    "leaflet.marker.slideto": "^0.2.0",
    "lodash": "^4.17.5",
    "mathjs": "^4.1.0",
    "moment": "^2.22.0",
    "simpl-schema": "^1.4.2",
    "sortablejs": "^1.7.0",
    "tinycolor2": "^1.4.1"
  }
}
```

**Comment by cjbassi** [comment]
I just updated/released a new version which adds more logging and fixes the error handling so it doesn't break your terminal on error. Could you run it again and let me know what it outputs this time? Thanks.

**Comment by chneau** [comment]
Well done, here is the output:

```
Error caught. Exiting program.

Issue with sparkline.

Diagnostics:
len(line.Data): 104
max: 31350991
x: 134
self.X: 134
offset: 0
cur_item: -80822124
percent: -2.5779766897958662
index: -18
len(SPARKS): 8

panic: 1

goroutine 159 [running]:
github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui.Error(0x58579d, 0x9, 0xc420dfe870, 0x89)
	/home/c/go/src/github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui/utils.go:35 +0x23f
github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui.(*Sparklines).Buffer(0xc4200b52a0, 0x0)
	/home/c/go/src/github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui/sparkline.go:79 +0x65b
github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui.(*Grid).Buffer(0xc4201e2000, 0x58e448)
	/home/c/go/src/github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui/grid.go:53 +0xbc
github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui.Render.func1(0xc4203a8000, 0x5a16e0, 0xc4201e2000)
	/home/c/go/src/github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui/render.go:23 +0x76
created by github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui.Render
	/home/c/go/src/github.com/cjbassi/gotop/vendor/github.com/cjbassi/termui/render.go:21 +0xa7
```

Using the same method, gotop did crash during a meteor build on a docker container.

**Comment by cjbassi** [comment]
Got one more update for you to try. The issue is that you're getting negative values for network usage, so that might be an issue with gopsutil. If we don't get any new info this time then I'm not sure what's going on and I'll just start ignoring negative values for the sparkline graph which should stop the error.

**Comment by chneau** [comment]
Here is the new trace:

```
10:07:49-c@c:~$gotop
Error caught. Exiting program.

Issue with net data.

Diagnostics:
curRecvTotal: 1136503439
curSentTotal: 172884658
self.prevRecvTotal: 1136900447
self.prevSentTotal: 253890700
recvRecent: 18446744073709154608
sentRecent: 18446744073628545574
int(recvRecent): -397008
int(sentRecent): -81006042

panic: 1

goroutine 16 [running]:
github.com/cjbassi/gotop/utils.Error(0x58521b, 0x8, 0xc4215985a0, 0xe3)
	/home/c/go/src/github.com/cjbassi/gotop/utils/utils.go:30 +0x23f
github.com/cjbassi/gotop/widgets.(*Net).update(0xc42000d320)
	/home/c/go/src/github.com/cjbassi/gotop/widgets/net.go:59 +0x516
github.com/cjbassi/gotop/widgets.NewNet.func1(0xc42008eaf0, 0xc42000d320)
	/home/c/go/src/github.com/cjbassi/gotop/widgets/net.go:38 +0x5c
created by github.com/cjbassi/gotop/widgets.NewNet
	/home/c/go/src/github.com/cjbassi/gotop/widgets/net.go:36 +0x222
10:09:55-c@c:~$
```

**Comment by cjbassi** [comment]
Supposedly fixed by c7171781c13f2bdd2f3cffd149800426a97dd4c2. Let me know if issues are still had.
